File: ruby/testcode/xxe_016.rb
require_relative 'shared'

module Nokogiri; module XML
  module ParseOptions
    NONET = 2048
    STRICT = 1
  end
  def self.parse(input, url = nil, encoding = nil, options = nil); input; end
end; end

# vuln-code-snippet start ruby_xxe_nokogiri_strict_opts
def parse_xml_strict(req)
  xml_input = req.body_str
  opts = Nokogiri::XML::ParseOptions::STRICT | Nokogiri::XML::ParseOptions::NONET
  doc = Nokogiri::XML.parse(xml_input, nil, nil, opts) # vuln-code-snippet safe-line ruby_xxe_nokogiri_strict_opts
  BenchmarkResponse.ok(doc.to_s)
end
# vuln-code-snippet end ruby_xxe_nokogiri_strict_opts

File: php/testcode/xxe_016.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_xxe_xslt_trusted_file
function xxe016(BenchmarkRequest $req): BenchmarkResponse {
    $data = $req->postData;
    $xml = new DOMDocument();
    $root = $xml->createElement('data');
    foreach ($data as $key => $val) {
        $child = $xml->createElement($key, htmlspecialchars($val, ENT_XML1));
        $root->appendChild($child);
    }
    $xml->appendChild($root);
    $xsl = new DOMDocument();
    $xsl->load(__DIR__ . '/templates/trusted_transform.xsl'); // vuln-code-snippet safe-line php_xxe_xslt_trusted_file
    $proc = new XSLTProcessor();
    $proc->importStylesheet($xsl);
    $output = $proc->transformToXml($xml);
    return BenchmarkResponse::ok($output);
}
// vuln-code-snippet end php_xxe_xslt_trusted_file
